### ISSUE TITLE
Question: Support for Transformer Engine

### ISSUE THREAD

**Issue by ConnorBaker** [patrick-kidger/equinox#669]
Hello! I really enjoy Equinox, especially the integrations with JaxTyping and your other projects.

I've been watching development on NVIDIA's Transformer Engine (https://github.com/NVIDIA/TransformerEngine) and I'm saddened by their support for frameworks built on JAX being limited to Flax and Praxis.

Looking through the code, there doesn't appear to be *a huge amount* in the way of Flax-specific functionality, so I have been considering trying to add support for Equinox. There are two things which caught my eye, though, and I was hoping for some insight prior to starting:
 
Usage of Flax variables and parameters: I'm unfamiliar with Flax, but it *seems* like variables are just PyTrees? Can they be replaced with actual attributes on the data class?

As an example, the base model from which all others inherit is defined here:

https://github.com/NVIDIA/TransformerEngine/blob/b8eea8aaa94bb566c3a12384eda064bda8ac4fd7/transformer_engine/jax/flax/module.py#L286-L333

Though that example leads to my second concern: axes with metadata? It seems like the code is making use of the Axis Metadata proposal (https://flax.readthedocs.io/en/latest/flip/2434-general-metadata.html#motivation) -- at least, that's what I'm gathering from the use of `variable_with_axes` and `param_with_axes` throughout. Any ideas on what refactoring to avoid using that feature might entail? Is there an equivalent feature in Equinox?

Any help would be greatly appreciated!

**Comment by dlwh** [comment]
Just to chime in as an outsider. We're adding support for TransformerEngine stuff to [Levanter](https://github.com/stanford-crfm/levanter/) (built with Equinox+NamedArrays). 

It's also worth mentioning that there's [a simpler FP8 thing in FLAX](https://github.com/google/flax/blob/main/flax/linen/fp8_ops.py) that I have been told is equally fast. As far as I'm aware, Equinox doesn't really have a blessed way of dealing with the OVERWRITE_WITH_GRADIENTS thing, but I think it's not too bad (pass around a filter tree and use apply_updates for one and overwrite with the other). My plan was to add an FP8Linear layer and just construct a filter tree based on that. The alternative is to do "dot injection" like in Flax.

The rest of TE is relatively straightforward to port I think?

**Comment by patrick-kidger** [comment]
Hey there! It'd be great to see support for Equinox+TransformerEnginer. This is actually something we discussed with NVIDIA a while back, and if I recall correctly it was something we were all interested in, and all that was required was someone with the time and interest to implement it!

As for your questions (although I should note I also don't tend to use Flax, so hopefully these answers are accurate...):
1. Indeed I believe the equivalence is Flax variables <-> Equinox annotations. Sometimes the Flax talk about mutating variables, in which case the appropriate thing would be [Equinox's stateful functionality](https://docs.kidger.site/equinox/api/nn/stateful/), more on this below.

2. For the FP8 stuff, then I basically like @dlwh's link -- this looks like a much simpler way to go about doing things.

3. In terms of the `OVERWRITE_WITH_GRADIENTS` that @dlwh has touched on, I think I have a simpler way of doing this, namely to use the stateful functionaliy mentioned above.
    
    First of all, if I understand what's going on correctly, this is a flag that Flax is using to state that the gradient of that parameter should in fact overwrite that parameter during backprop+SGD, rather than just adjusting it slightly.
    
    Now interestingly, the same design question came up in Equinox when designing `eqx.nn.BatchNorm`: how should we track the moving statistics? One suggestion was to "hide" the updated statistic in the gradient (that seems to be what they're doing here), and give Equinox some special notion of overwriting a value with its gradient, rather than adjusting a value with its gradient. But that clearly felt hackish -- it's tightly coupled to the notion of backpropagation, so how should this work with `jax.jvp`, which does forward-mode, or `jax.jacrev`, which does a batch-of-backprops??
    
    The solution was to realise that this was really about wanting a stateful operation: a means to preserve additional information from previous-output into next-input. And importantly that extra information doesn't (shouldn't!) have to live in the model PyTree. Thus the answer is actually to use the stateful functionality I've linked above. (Which is really just doing state in the way that every functional programming language since the dawn of time has always done state: have it as an extra input and output. Thus we get `(input, state) -> (output, state)` as a pure function -- not `input -> output` whilst quietly mutating some state somewhere as a side-effect.)

Bringing points 2. and 3. together, I think what this means is that the equivalent idiomatic solution in Equinox would be to:
- have a `jax.custom_jvp` for the matmul, so that the derivative pass is done in higher precision. (Just like @dlwh's link.)
- use `eqx.nn.{make_with_state, StateIndex}` to handle tracking the updating statistics. (Rather than the `custom_vjp` that is used in @dlwh's link.)

Gosh, that turned into a long message. Does that all make sense?

**Comment by dlwh** [comment]
I don't think that works here, unless I'm misunderstanding. A key difference between BatchNorm and FP8 is that the state with FP8 *depends on the gradient signal* (`out_qdq_bwd` calls `qdq_and_return` with `g` and updates the scale with `g`).

**Comment by patrick-kidger** [comment]
Ah, good point!

I suppose one could still do it (call `state = state.set(...)` outside the `grad` rather than inside the layer definition), but at that point it's very similar to what you were originally describing. (It really just depends on whether you prefer to hold the statistics in the model pytree or in a separate pytree, I guess.)

FWIW, if you see a technically feasible approach then I'd be happy to consider adjusting Equinox to make this use-case easier. (I.e. so that the updates are local to the model definition. Maybe we can adjust how `eqx.nn.State` works somehow?)
